presto-main/src/main/java/com/facebook/presto/operator/exchange/LocalExchange.java

```diff
@@ -0,0 +1,584 @@
+        {
+            return bufferedPages;
+        }
+
+        @Override
+        public LocalExchangeBufferInfo mergeWith(LocalExchangeBufferInfo other)
+        {
+            return new LocalExchangeBufferInfo(bufferedBytes + other.getBufferedBytes(), bufferedPages + other.getBufferedPages());
+        }
+    }
+
+    @ThreadSafe
+    public static class LocalExchangeSinkFactory
+            implements Closeable
+    {
+        private final LocalExchange exchange;
+
+        @GuardedBy("this")
+        private boolean closed;
+
+        private LocalExchangeSinkFactory(LocalExchange exchange)
+        {
+            this.exchange = exchange;
+        }
+
+        public List<Type> getTypes()
+        {
+            return exchange.getTypes();
+        }
+
+        public LocalExchangeSink createSink()
+        {
+            checkNotHoldsLock(this);
+
+            synchronized (this) {
+                checkState(!closed, "Factory already closed");
+            }
+            return exchange.createSink(this);
+        }
+
+        public LocalExchangeSinkFactory duplicate()
+        {
+            checkNotHoldsLock(this);
+
+            synchronized (this) {
+                checkState(!closed, "Factory already closed");
+            }
+            return exchange.createSinkFactory();
+        }
+
+        @Override
+        public void close()
+        {
+            checkNotHoldsLock(this);
+
+            synchronized (this) {
+                if (closed) {
+                    return;
+                }
+                closed = true;
+            }
+            exchange.sinkFactoryClosed(this);
+        }
+
+        public void noMoreSinkFactories()
+        {
+            checkNotHoldsLock(this);
+
+            exchange.noMoreSinkFactories();
+        }
+    }
+
+    @ThreadSafe
+    public static class LocalExchangeSink
+    {
+        private final LocalExchange exchange;
+        private final Consumer<Page> sink;
+
+        @GuardedBy("this")
+        private boolean finished;
+
+        private LocalExchangeSink(LocalExchange exchange, Consumer<Page> sink)
+        {
+            this.exchange = exchange;
+            this.sink = sink;
+        }
+
+        public List<Type> getTypes()
+        {
+            return exchange.getTypes();
+        }
+
+        public void finished()
+        {
+            checkNotHoldsLock(this);
+
+            synchronized (this) {
+                if (finished) {
+                    return;
+                }
+                finished = true;
+            }
+
+            // Notify exchange outside of lock
+            exchange.sinkFinished(this);
+        }
+
+        public boolean isFinished()
+        {
+            checkNotHoldsLock(this);
+
+            synchronized (this) {
+                return finished;
+            }
+        }
+
+        public ListenableFuture<?> waitForWriting()
+        {
+            checkNotHoldsLock(this);
+
+            if (isFinished()) {
```

Comment:
The guarantee provided by this check seems counter-intuitive. While this check makes sure that new threads that try to write will not block at all, existing threads that are blocked at the time `finish` is called are not unblocked.